03_Modules/Transactions/src/module/services/transaction.event.service.ts

```diff
@@ -0,0 +1,530 @@
+        );
+        if (
+          overallValue &&
+          overallValue.unitOfMeasure?.unit?.toUpperCase() === "KWH"
+        ) {
+          valuesMap.set(Date.parse(meterValue.timestamp), overallValue.value);
+        } else if (
+          overallValue &&
+          overallValue.unitOfMeasure?.unit?.toUpperCase() === "WH"
+        ) {
+          valuesMap.set(
+            Date.parse(meterValue.timestamp),
+            overallValue.value / 1000
+          );
+        }
+      });
+
+    // sort the map based on timestamps
+    valuesMap = new Map(
+      [...valuesMap.entries()].sort((v1, v2) => v1[0] - v2[0])
+    );
+    const sortedValues = Array.from(valuesMap.values());
+
+    let totalKwh: number = 0;
+    for (let i = 1; i < sortedValues.length; i++) {
+      totalKwh += sortedValues[i] - sortedValues[i - 1];
+    }
+
+    return totalKwh;
+  }
+
+  /**
+   * Internal method to execute a costUpdated request for an ongoing transaction repeatedly based on the costUpdatedInterval
+   *
+   * @param {string} stationId - The identifier of the client connection.
+   * @param {string} transactionId - The identifier of the transaction.
+   * @param {number} costUpdatedInterval - The costUpdated interval in milliseconds.
+   * @param {string} tenantId - The identifier of the tenant.
+   * @return {void} This function does not return anything.
+   */
+  private updateCost(
+    stationId: string,
+    transactionId: string,
+    costUpdatedInterval: number,
+    tenantId: string
+  ): void {
+    setInterval(async () => {
+      const transaction: Transaction | undefined =
+        await this.transactionEventRepository?.readTransactionByStationIdAndTransactionId(
+          stationId,
+          transactionId
+        );
+      if (transaction && transaction.isActive) {
+        const cost = await this._calculateTotalCost(stationId, transaction.id);
+        this.sendCall(stationId, tenantId, CallAction.CostUpdated, {
+          totalCost: cost,
+          transactionId: transaction.transactionId,
+        } as CostUpdatedRequest).then(() => {
+          this.loggerService?.logger.info(
+            `Sent costUpdated for ${transaction.transactionId} with totalCost ${cost}`
+          );
+        });
+      }
+    }, costUpdatedInterval * 1000);
+  }
+
+  // todo move to base service?
+  /**
+   * Sends a call with the specified identifier, tenantId, action, payload, and origin.
+   *
+   * @param {string} identifier - The identifier of the call.
+   * @param {string} tenantId - The tenant ID.
+   * @param {CallAction} action - The action to be performed.
+   * @param {OcppRequest} payload - The payload of the call.
+   * @param {string} [callbackUrl] - The callback URL for the call.
+   * @param {string} [correlationId] - The correlation ID of the call.
+   * @param {MessageOrigin} [origin] - The origin of the call.
+   * @return {Promise<IMessageConfirmation>} A promise that resolves to the message confirmation.
+   */
+  public sendCall(
```

Comment:
This comes from base, since this simply involves the cacheService and the sender (along with logger), I would say that it makes sense for this to move as we utilize DI more to adjust our structure to be service based. We have a lot of options of how this method can be shared, it can be in some Base class, it can be injected, it can be a static method on a class, but I think having this in Base makes sense since the modules had this in base